### PR TITLE
Transpose Operator for GPU

### PR DESCRIPTION
**Pull request by Kh4L**
Refactor of #421. Separating original cuTT code and DALI modifications.

Efficient **Transpose** operator on GPU with [cuTT](https://arxiv.org/abs/1705.01598)'s original implementation.

First commit is a code drop of [cuTT](https://github.com/ap-hynninen/cutt) in `dali/pipeline/operators/transpose/cutt`.

### CODE REVIEW

**Comment by Kh4L** [comment]
Build 624493

File: dali/pipeline/operators/transpose/transpose.cu [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+  for (int i = 0; i < rank; ++i) {
+    new_dims[i] = dims[rank - 1 - i];
+    new_perm[i] = rank - 1 - perm[rank - 1 - i];
+  }
+  return {std::move(new_dims), std::move(new_perm)};
+}
+
+template <typename T = double>
+void cuttSanityCheck(int *dims, int *perm, int rank) {
```
Comment (klecki) [inline dali/pipeline/operators/transpose/transpose.cu]: Remove this as well.

File: dali/pipeline/operators/transpose/transpose_test.cc [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+std::vector<testing::Arguments> permutationsRank3 = {
+    {{"perm", std::vector<int>{0, 1, 2}}},
+    {{"perm", std::vector<int>{1, 2, 0}}},
+    {{"perm", std::vector<int>{2, 1, 0}}},
+    {{"perm", std::vector<int>{2, 0, 1}}},
+    {{"perm", std::vector<int>{0, 2, 1}}},
+    {{"perm", std::vector<int>{1, 0, 2}}},
+};
+
+std::vector<testing::Arguments> permutationsRank4 = {
```
Comment (klecki) [inline dali/pipeline/operators/transpose/transpose_test.cc]: I would be lazy here: https://en.cppreference.com/w/cpp/algorithm/next_permutation

File: dali/pipeline/operators/transpose/transpose_test.cc [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+                        const std::vector<Index>& new_shape,
+                        const std::vector<int>& perm) {
+  auto old_volume = Volume(old_shape);
+  auto new_volume = Volume(new_shape);
+  ASSERT_EQ(old_volume, new_volume);
+
+
+  auto old_strides = GetStrides(old_shape);
+  auto new_strides = GetStrides(new_shape);
+  if (new_shape.size() == 3) {
```
Comment (klecki) [inline dali/pipeline/operators/transpose/transpose_test.cc]: Why not generic?

File: dali/pipeline/operators/transpose/transpose_test.cc [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+namespace dali {
+
+namespace {
+
+// Fill tensors of consecutive numbers
+template <typename T>
+void Arrange(T* ptr, const std::vector<Index>& shape) {
+  auto volume = Volume(shape);
+  for (int i = 0; i < volume; ++i) {
+    ptr[i] = static_cast<int>(i);
```
[the remark this answers is not in the document]
Comment (Kh4L) [reply dali/pipeline/operators/transpose/transpose_test.cc]: It was :+1:

File: dali/pipeline/operators/transpose/cutt/CudaMemcpy.cu [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+  for (int i = threadIdx.x + blockIdx.x*blockDim.x;i < n;i += blockDim.x*gridDim.x) {
+    data_out[i] = data_in[i];
+  }
+
+}
+template <typename T>
+void scalarCopy(const int n, const T* data_in, T* data_out, cudaStream_t stream) {
+
+  int numblock = (n - 1)/numthread + 1;
+  // numblock = min(65535, numblock);
```
Comment (JanuszL) [inline dali/pipeline/operators/transpose/cutt/CudaMemcpy.cu]: ?

File: dali/pipeline/operators/transpose/transpose.cc [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+#include "dali/pipeline/operators/transpose/transpose.h"
+
+namespace dali {
+
+DALI_SCHEMA(Transpose)
+  .DocStr("Transpose tensor dimension to a new permutated dimension specified by `perm`.")
+  .NumInput(1)
+  .NumOutput(1)
+  // .AllowMultipleInputSets()
```
Comment (JanuszL) [inline dali/pipeline/operators/transpose/transpose.cc]: ?

File: dali/pipeline/operators/transpose/transpose_test.cc (hunk 3 above)
Comment (Kh4L) [reply dali/pipeline/operators/transpose/transpose_test.cc]: Done

File: dali/pipeline/operators/transpose/cutt/CudaMemcpy.cu [hunk 7]
```diff
@@ -0,0 +1,153 @@
+/******************************************************************************
+MIT License
+
+Copyright (c) 2016 Antti-Pekka Hynninen
+Copyright (c) 2016 Oak Ridge National Laboratory (UT-Batelle)
```
Comment (jantonguirao) [inline dali/pipeline/operators/transpose/cutt/CudaMemcpy.cu]: how about our copyright notice? at least you changed the include paths

Comment (Kh4L) [reply dali/pipeline/operators/transpose/cutt/CudaMemcpy.cu]: I add ours only to the modified files
So I'll add ours to this file

Comment (Kh4L) [reply dali/pipeline/operators/transpose/cutt/CudaMemcpy.cu]: Done

**Comment by Kh4L** [comment]
Build 630454

**Pull request merged**